### ISSUE TITLE
Best way to handle environment specific configuration

### ISSUE THREAD

**Issue by TeaBough** [CiscoCloud/microservices-infrastructure#924]
How do you handle environment specific configuration for your apps (db passwords, host/port of underlying apps ...) ? I've some solutions but none of them is completely satisfying to me. Here is what I've managed to come up with : 

**1) Deploying all of my configuration files on every Mesos slaves**
In this case I use a tool like Ansible/Chef/Puppet to push my configurations file on every Mesos slaves.

Pros : 
* I inject the configuration using volumes, thus I don't need to change my application image to handle the configuration

Cons : 
* Everytime I change a configuration file I need to redeploy it on every host 
* If I want more than one version of my application with different config to run on the same host I need to add folder to add namespace ( v0.1/<MY_APP_NAME>/application.properties)
* There might be security issues ( my db password would be on every slaves in a file ...)

**2) Add envconsul to every container**

I add `envconsul` to every container allowing me to wrap the start command like this : 
`envconsul -consul consul.service.consul:8500 -once -prefix marathon/my-app my-app.sh`

My app can now access env vars that are synced with my consul K/V store.

Pros : 
* My configuration resides only on Consul and the container that need it
* I can implement dynamic runtime configuration

Cons : 
* I need to add envconsul to every container which goes against the core principle of Microservices and Containers which is the separation of concerns

**3) Generate a custom Marathon deployment json file**

I use a tool that will get values from Consul K/V and generate a Marathon deployment file with the correct env vars set.

Pros : 
* I inject the configuration env variables directly in the json file, so no need to change the image

Cons :
* To my knowledge there are no such tool that will create for you a deploy file with env var pulled from Consul

**4) Wrap `docker run` with envconsul**
I do something like that in my marathon file : 
`envconsul -consul consul.service.consul:8500 -once -prefix marathon/my-app docker run my-app`

Pros : 
* I inject the configuration env variables directly in the json file

Cons :
* I need to develop a wrapper that will create my custom docker run command based on the K/V consul
* I can't use anymore the fact that Marathon/Mesos uses containers as first class citizen.

**5) Using a sidecar/sidekick that containes envconsul**
The sidecar container (containing my configuration) will be mounted in the container app as a volume-from.

This is actually not possible with Marathon since there isn't the notion of Pods like in K8s that assure you that 2 containers will run on the same host.

**Comment by stevendborrelli** [comment]
We'd like to use vault for this. We should have some engineering resources dedicated to this starting next week. 

What we are thinking right now;
1. Start with a vault docker plugin. This means when you launch a container, you send a token with mounting instructions. 
2. The vault plugin pulls the secret from the backend, and mounts it into the container using a tmpfs/fuse on the host.

**Comment by eirslett** [comment]
You could just put your marathon.json files in a git repository, coupled with their configuration (which is then not stored in consul k/v, but stored in git).
For secrets, vault is probably a good fit.

**Comment by TeaBough** [comment]
@stevendborrelli Will you  handle non-secret environment configuration with Vault  ? I don't understand how you plan to send mounting instructions.
@eirslett Yeah but using the K/V store Consul is nice in order to handle ACL, and it allows more flexibility that a github repo.

**Comment by TeaBough** [comment]
@stevendborrelli & @eirslett  What do you think of solution solution 3) ?

**Comment by eirslett** [comment]
We've not had the need for using consul's ACL feature in the K/V store, at least not at the application level. In any case, the "tool" that would convert consul data to marathon.json would have access to read anything from consul anyways, so that it could do its job.
And consul k/v isn't really more flexible; git comes with a lot of pros, like traceability (who did what, when), availability (as long as somebody has a copy of the repo, you won't lose the data), visibility - it's all there as files, and scriptability - it's easier to script changes to static files than to data you have to read and write from an API. (We found git is actually more flexible)

**Comment by TeaBough** [comment]
We need to have teams being able to modify and handle the different environments configurations. But in the same time we don't want them to being handle to change other teams configuration and we don't want them to access sensible variables (DB passwords ...). Having a central git repository with all the configuration of all our services would mean we won't be able to open it to every teams since they will be able to modify configuration of other teams. If we create one configuration repo by service it would mean duplication of certain configurations. For instance if the port of a service changes we need to change it on every git repo that needs it.

**Comment by eirslett** [comment]
Yes, if it's going to be feasible with git, all configuration has to be in the same repo. And all developers must have access to it. We generally trust our developers not to do things they're not supposed to. (They often have a valid use case to change the configuration of another project, e.g. a consumer of their microservice)
Of course, you shouldn't have secrets commited in git (unless they're encrypted, in that case it's fine) so that has to be solved in another way.

**Comment by TeaBough** [comment]
We trust as well our developers but we try to limit the impact of an error a team or a dev can make.

**Comment by stevendborrelli** [comment]
Have you folks seen git2consul? https://github.com/Cimpress-MCP/git2consul. I've not tried it myself, but it looks like an interesting solution. 

Marathon has the capability to download files into the mesos sandbox using URIs. That's another option to get config values into containers.

Having a k/v store and vault included in the cluster gives us a lot of options, but we haven't nailed down this experience yet.

For 0.6, we're going to merge the ACL enforcement on consul: https://github.com/CiscoCloud/microservices-infrastructure/pull/603

We're also trying to get vault 0.4 in the next release (mid-January). 

We're also working on a tool called gestalt: https://github.com/asteris-llc/gestalt, which we envision as a way to make it easy for developers to store schema enforced configurations in a k/v store.

**Comment by crumley** [comment]
Where did this end up? Anyone end up with a solution they are happy with?

**Comment by siddharthist** [comment]
I'd like to add another option, which is 

**5) Solution 1) + consul-template**

That way, you can only deploy the config file template once, update the values dynamically in Consul Template, and have Consul Template tell Marathon to do a rolling restart of your app.

**Comment by crumley** [comment]
5 + 1 is nice. Simple and doesn't require changes to the marathon applications beyond adding a `load from external configuration file` config.
